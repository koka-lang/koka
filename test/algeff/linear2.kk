
effect state<s> {
  fun get() : s  
  fun put(x : s) : ()
}

val state = handler(s) {
  return x -> x
  get()    -> resume(s,s)
  put(t)   -> 20 + resume((),t) + resume((),t)
}

fun foo() {
  val i = get()
  put(i+1)
  put(i+2)
  get()
}

fun main() {
  state(0,foo).println  // 68
}