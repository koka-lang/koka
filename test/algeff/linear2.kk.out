[01;30moperator branch (state) get: resume tail
[00;39;49m[00m[01;30moperator branch (state) put: resume scoped
[00;39;49m[00m68
.Op-get : forall<a> .op-get<a>
.Op-put : forall<a> (x : a) -> .op-put<a>
.Ops-get : forall<a,b> (get : .op-get<a>) -> .ops-state<a,a>
.Ops-put : forall<a,b> (put : .op-put<a>) -> .ops-state<a,()>
.tag-state : string
foo : () -> (state<int>) int
get : forall<a> () -> (state<a>) a
main : () -> console ()
put : forall<a> (x : a) -> (state<a>) ()
state : forall<a,e> (s : a, action : () -> <state<a>|e> int) -> e int